## Linear Regression with PyTorch

#### Part 2 of "PyTorch: Zero to GANs"

*This post is the second in a series of tutorials on building deep learning models with PyTorch, an open source neural networks library developed and maintained by Facebook. Check out the full series:*


1. [PyTorch Basics: Tensors & Gradients](https://jovian.ml/aakashns/01-pytorch-basics)
2. [Linear Regression & Gradient Descent](https://jovian.ml/aakashns/02-linear-regression)
3. [Image Classfication using Logistic Regression](https://jovian.ml/aakashns/03-logistic-regression) 
4. [Training Deep Neural Networks on a GPU](https://jovian.ml/aakashns/04-feedforward-nn)
5. [Image Classification using Convolutional Neural Networks](https://jovian.ml/aakashns/05-cifar10-cnn)
6. [Data Augmentation, Regularization and ResNets](https://jovian.ml/aakashns/05b-cifar10-resnet)
7. [Generating Images using Generative Adverserial Networks](https://jovian.ml/aakashns/06-mnist-gan)

Continuing where the [previous tutorial](https://jvn.io/aakashns/3143ceb92b4f4cbbb4f30e203580b77b) left off, we'll discuss one of the foundational  algorithms of machine learning in this post: *Linear regression*. We'll create a model that predicts crop yields for apples and oranges (*target variables*) by looking at the average temperature, rainfall and humidity (*input variables or features*) in a region. Here's the training data:

![linear-regression-training-data](https://i.imgur.com/6Ujttb4.png)

In a linear regression model, each target variable is estimated to be a weighted sum of the input variables, offset by some constant, known as a bias :

```
yield_apple  = w11 * temp + w12 * rainfall + w13 * humidity + b1
yield_orange = w21 * temp + w22 * rainfall + w23 * humidity + b2
```

Visually, it means that the yield of apples is a linear or planar function of temperature, rainfall and humidity:

![linear-regression-graph](https://i.imgur.com/4DJ9f8X.png)

The *learning* part of linear regression is to figure out a set of weights `w11, w12,... w23, b1 & b2` by looking at the training data, to make accurate predictions for new data (i.e. to predict the yields for apples and oranges in a new region using the average temperature, rainfall and humidity). This is done by adjusting the weights slightly many times to make better predictions, using an optimization technique called *gradient descent*.

## System setup

This tutorial takes a code-first approach towards learning PyTorch, and you should try to follow along by running and experimenting with the code yourself. The easiest way to start executing this notebook is to click the **"Run"** button at the top of this page, and select **"Run on Binder"**. This will run the notebook on [mybinder.org](https://mybinder.org), a free online service for running Jupyter notebooks.

**NOTE**: *If you're running this notebook on Binder, please skip ahead to the next section.*

### Running on your computer locally

You can clone this notebook hosted on [Jovian.ml](https://www.jovian.ml), install the required dependencies, and start Jupyter by running the following commands on the terminal:

```bash
pip install jovian --upgrade    # Install the jovian library 
jovian clone aakashns/02-linear-regression      # Download notebook & dependencies
cd 02-linear-regression         # Enter the created directory 
jovian install                  # Install the dependencies
conda activate 02-linear-regression # Activate virtual environment
jupyter notebook                # Start Jupyter
```

On older versions of conda, you might need to run `source activate 02-linear-regression` to activate the environment. For a more detailed explanation of the above steps, check out the *System setup* section in the [previous notebook](https://jovian.ml/aakashns/01-pytorch-basics).

We begin by importing Numpy and PyTorch:

In [1]:
# Uncomment the command below if Numpy or PyTorch is not installed
# !conda install numpy pytorch cpuonly -c pytorch -y

In [2]:
import numpy as np
import torch

## Training data

The training data can be represented using 2 matrices: `inputs` and `targets`, each with one row per observation, and one column per variable.

In [3]:
# Input (temp, rainfall, humidity)
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70]], dtype='float32')

In [4]:
# Targets (apples, oranges)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119]], dtype='float32')

We've separated the input and target variables, because we'll operate on them separately. Also, we've created numpy arrays, because this is typically how you would work with training data: read some CSV files as numpy arrays, do some processing, and then convert them to PyTorch tensors as follows:

In [5]:
# Convert inputs and targets to tensors
inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)
print(inputs)
print(targets)

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.]])
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


## Linear regression model from scratch

The weights and biases (`w11, w12,... w23, b1 & b2`) can also be represented as matrices, initialized as random values. The first row of `w` and the first element of `b` are used to predict the first target variable i.e. yield of apples, and similarly the second for oranges.

In [6]:
# Weights and biases
w = torch.randn(2, 3, requires_grad=True)
b = torch.randn(2, requires_grad=True)
print(w)
print(b)

tensor([[-0.0313,  0.3479, -0.2870],
        [ 0.0531, -0.4948, -0.9435]], requires_grad=True)
tensor([-0.6292, -0.7645], requires_grad=True)


`torch.randn` creates a tensor with the given shape, with elements picked randomly from a [normal distribution](https://en.wikipedia.org/wiki/Normal_distribution) with mean 0 and standard deviation 1.

Our *model* is simply a function that performs a matrix multiplication of the `inputs` and the weights `w` (transposed) and adds the bias `b` (replicated for each observation).

![matrix-mult](https://i.imgur.com/WGXLFvA.png)

We can define the model as follows:

In [7]:
def model(x):
    return x @ w.t() + b

`@` represents matrix multiplication in PyTorch, and the `.t` method returns the transpose of a tensor.

The matrix obtained by passing the input data into the model is a set of predictions for the target variables.

In [8]:
# Generate predictions
preds = model(inputs)
print(preds)

tensor([[   8.0487,  -70.6085],
        [   8.7626,  -99.8566],
        [  26.6120, -117.1691],
        [   0.5132,  -51.5316],
        [  10.5129, -110.6448]], grad_fn=<AddBackward0>)


Let's compare the predictions of our model with the actual targets.

In [9]:
# Compare with targets
print(targets)

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


You can see that there's a huge difference between the predictions of our model, and the actual values of the target variables. Obviously, this is because we've initialized our model with random weights and biases, and we can't expect it to *just work*.

## Loss function

Before we improve our model, we need a way to evaluate how well our model is performing. We can compare the model's predictions with the actual targets, using the following method:

* Calculate the difference between the two matrices (`preds` and `targets`).
* Square all elements of the difference matrix to remove negative values.
* Calculate the average of the elements in the resulting matrix.

The result is a single number, known as the **mean squared error** (MSE).

In [10]:
# MSE loss
def mse(t1, t2):
    diff = t1 - t2
    return torch.sum(diff * diff) / diff.numel()

`torch.sum` returns the sum of all the elements in a tensor, and the `.numel` method returns the number of elements in a tensor. Let's compute the mean squared error for the current predictions of our model.

In [11]:
# Compute loss
loss = mse(preds, targets)
print(loss)

tensor(20834.1953, grad_fn=<DivBackward0>)


Here’s how we can interpret the result: *On average, each element in the prediction differs from the actual target by about 145 (square root of the loss 20834)*. And that’s pretty bad, considering the numbers we are trying to predict are themselves in the range 50–200. Also, the result is called the *loss*, because it indicates how bad the model is at predicting the target variables. Lower the loss, better the model.

## Compute gradients

With PyTorch, we can automatically compute the gradient or derivative of the loss w.r.t. to the weights and biases, because they have `requires_grad` set to `True`.

In [12]:
# Compute gradients
loss.backward()

The gradients are stored in the `.grad` property of the respective tensors. Note that the derivative of the loss w.r.t. the weights matrix is itself a matrix, with the same dimensions.

In [13]:
# Gradients for weights
print(w)
print(w.grad)

tensor([[-0.0313,  0.3479, -0.2870],
        [ 0.0531, -0.4948, -0.9435]], requires_grad=True)
tensor([[ -5337.0156,  -6350.4648,  -3862.5439],
        [-15036.5605, -17294.3145, -10552.3203]])


The loss is a [quadratic function](https://en.wikipedia.org/wiki/Quadratic_function) of our weights and biases, and our objective is to find the set of weights where the loss is the lowest. If we plot a graph of the loss w.r.t any individual weight or bias element, it will look like the figure shown below. A key insight from calculus is that the gradient indicates the rate of change of the loss, or the [slope](https://en.wikipedia.org/wiki/Slope) of the loss function w.r.t. the weights and biases.

If a gradient element is **positive**:
* **increasing** the element's value slightly will **increase** the loss.
* **decreasing** the element's value slightly will **decrease** the loss

![postive-gradient](https://i.imgur.com/hFYoVgU.png)

If a gradient element is **negative**:
* **increasing** the element's value slightly will **decrease** the loss.
* **decreasing** the element's value slightly will **increase** the loss.

![negative=gradient](https://i.imgur.com/w3Wii7C.png)

The increase or decrease in loss by changing a weight element is proportional to the value of the gradient of the loss w.r.t. that element. This forms the basis for the optimization algorithm that we'll use to improve our model.

Before we proceed, we reset the gradients to zero by calling `.zero_()` method. We need to do this, because PyTorch accumulates, gradients i.e. the next time we call `.backward` on the loss, the new gradient values will get added to the existing gradient values, which may lead to unexpected results.

In [14]:
w.grad.zero_()
b.grad.zero_()
print(w.grad)
print(b.grad)

tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([0., 0.])


## Adjust weights and biases using gradient descent

We'll reduce the loss and improve our model using the gradient descent optimization algorithm, which has the following steps:

1. Generate predictions

2. Calculate the loss

3. Compute gradients w.r.t the weights and biases

4. Adjust the weights by subtracting a small quantity proportional to the gradient

5. Reset the gradients to zero

Let's implement the above step by step.

In [15]:
# Generate predictions
preds = model(inputs)
print(preds)

tensor([[   8.0487,  -70.6085],
        [   8.7626,  -99.8566],
        [  26.6120, -117.1691],
        [   0.5132,  -51.5316],
        [  10.5129, -110.6448]], grad_fn=<AddBackward0>)


Note that the predictions are same as before, since we haven't made any changes to our model. The same holds true for the loss and gradients.

In [16]:
# Calculate the loss
loss = mse(preds, targets)
print(loss)

tensor(20834.1953, grad_fn=<DivBackward0>)


In [17]:
# Compute gradients
loss.backward()
print(w.grad)
print(b.grad)

tensor([[ -5337.0156,  -6350.4648,  -3862.5439],
        [-15036.5605, -17294.3145, -10552.3203]])
tensor([ -65.3101, -181.9621])


Finally, we update the weights and biases using the gradients computed above.

In [18]:
# Adjust weights & reset gradients
with torch.no_grad():
    w -= w.grad * 1e-5
    b -= b.grad * 1e-5
    w.grad.zero_()
    b.grad.zero_()

A few things to note above:

* We use `torch.no_grad` to indicate to PyTorch that we shouldn't track, calculate or modify gradients while updating the weights and biases. 

* We multiply the gradients with a really small number (`10^-5` in this case), to ensure that we don't modify the weights by a really large amount, since we only want to take a small step in the downhill direction of the gradient. This number is called the *learning rate* of the algorithm. 

* After we have updated the weights, we reset the gradients back to zero, to avoid affecting any future computations.

Let's take a look at the new weights and biases.

In [19]:
print(w)
print(b)

tensor([[ 0.0220,  0.4114, -0.2484],
        [ 0.2035, -0.3219, -0.8380]], requires_grad=True)
tensor([-0.6285, -0.7626], requires_grad=True)


With the new weights and biases, the model should have lower loss.

In [20]:
# Calculate loss
preds = model(inputs)
loss = mse(preds, targets)
print(loss)

tensor(14275.3809, grad_fn=<DivBackward0>)


We have already achieved a significant reduction in the loss, simply by adjusting the weights and biases slightly using gradient descent.

## Train for multiple epochs

To reduce the loss further, we can repeat the process of adjusting the weights and biases using the gradients multiple times. Each iteration is called an epoch. Let's train the model for 100 epochs.

In [21]:
# Train for 100 epochs
for i in range(100):
    preds = model(inputs)
    loss = mse(preds, targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

Once again, let's verify that the loss is now lower:

In [22]:
# Calculate loss
preds = model(inputs)
loss = mse(preds, targets)
print(loss)

tensor(267.1306, grad_fn=<DivBackward0>)


As you can see, the loss is now much lower than what we started out with. Let's look at the model's predictions and compare them with the targets.

In [23]:
# Predictions
preds

tensor([[ 60.4480,  76.3242],
        [ 79.5420,  96.6628],
        [119.4743, 132.3997],
        [ 40.2797,  71.8990],
        [ 85.9573,  91.6603]], grad_fn=<AddBackward0>)

In [24]:
# Targets
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])

The prediction are now quite close to the target variables, and we can get even better results by training for a few more epochs. 

At this point, we can save our notebook and upload it to [Jovian.ml](https://www.jovian.ml) for future reference and sharing.

In [ ]:
!pip install jovian --upgrade -q

In [27]:
import jovian

In [28]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Please enter your API key ( from https://jovian.ml/ ):
API KEY: ········
[jovian] Updating notebook "aakashns/02-linear-regression" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/aakashns/02-linear-regression


'https://jovian.ml/aakashns/02-linear-regression'

`jovian.commit` uploads the notebook to [Jovian.ml](https://www.jovian.ml), captures the Python environment and creates a sharable link for the notebook. You can use this link to share your work and let anyone reproduce it easily with the `jovian clone` command.  Jovian also includes a powerful commenting interface, so you (and others) can discuss & comment on specific parts of your notebook:

![commenting on jovian](https://cdn-images-1.medium.com/max/1600/1*b4snnr_5Ve5Nyq60iDtuuw.png)

## Linear regression using PyTorch built-ins

The model and training process above were implemented using basic matrix operations. But since this such a common pattern , PyTorch has several built-in functions and classes to make it easy to create and train models.

Let's begin by importing the `torch.nn` package from PyTorch, which contains utility classes for building neural networks.

In [29]:
import torch.nn as nn

As before, we represent the inputs and targets and matrices.

In [30]:
# Input (temp, rainfall, humidity)
inputs = np.array([[73, 67, 43], [91, 88, 64], [87, 134, 58], 
                   [102, 43, 37], [69, 96, 70], [73, 67, 43], 
                   [91, 88, 64], [87, 134, 58], [102, 43, 37], 
                   [69, 96, 70], [73, 67, 43], [91, 88, 64], 
                   [87, 134, 58], [102, 43, 37], [69, 96, 70]], 
                  dtype='float32')

# Targets (apples, oranges)
targets = np.array([[56, 70], [81, 101], [119, 133], 
                    [22, 37], [103, 119], [56, 70], 
                    [81, 101], [119, 133], [22, 37], 
                    [103, 119], [56, 70], [81, 101], 
                    [119, 133], [22, 37], [103, 119]], 
                   dtype='float32')

inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)

We are using 15 training examples this time, to illustrate how to work with large datasets in small batches.

## Dataset and DataLoader

We'll create a `TensorDataset`, which allows access to rows from `inputs` and `targets` as tuples, and provides standard APIs for working with many different types of datasets in PyTorch.

In [31]:
from torch.utils.data import TensorDataset

In [32]:
# Define dataset
train_ds = TensorDataset(inputs, targets)
train_ds[0:3]

(tensor([[ 73.,  67.,  43.],
         [ 91.,  88.,  64.],
         [ 87., 134.,  58.]]), tensor([[ 56.,  70.],
         [ 81., 101.],
         [119., 133.]]))

The `TensorDataset` allows us to access a small section of the training data using the array indexing notation (`[0:3]` in the above code). It returns a tuple (or pair), in which the first element contains the input variables for the selected rows, and the second contains the targets.

We'll also create a `DataLoader`, which can split the data into batches of a predefined size while training. It also provides other utilities like shuffling and random sampling of the data.

In [33]:
from torch.utils.data import DataLoader

In [34]:
# Define data loader
batch_size = 5
train_dl = DataLoader(train_ds, batch_size, shuffle=True)

The data loader is typically used in a `for-in` loop. Let's look at an example.

In [35]:
for xb, yb in train_dl:
    print(xb)
    print(yb)
    break

tensor([[ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 91.,  88.,  64.],
        [ 73.,  67.,  43.],
        [ 69.,  96.,  70.]])
tensor([[119., 133.],
        [ 22.,  37.],
        [ 81., 101.],
        [ 56.,  70.],
        [103., 119.]])


In each iteration, the data loader returns one batch of data, with the given batch size. If `shuffle` is set to `True`, it shuffles the training data before creating batches. Shuffling helps randomize the input to the optimization algorithm, which can lead to faster reduction in the loss.

## nn.Linear

Instead of initializing the weights & biases manually, we can define the model using the `nn.Linear` class from PyTorch, which does it automatically.

In [36]:
# Define model
model = nn.Linear(3, 2)
print(model.weight)
print(model.bias)

Parameter containing:
tensor([[ 0.0933,  0.5745, -0.2043],
        [ 0.3817, -0.3547,  0.0262]], requires_grad=True)
Parameter containing:
tensor([-0.4140,  0.1716], requires_grad=True)


PyTorch models also have a helpful `.parameters` method, which returns a list containing all the weights and bias matrices present in the model. For our linear regression model, we have one weight matrix and one bias matrix.

In [37]:
# Parameters
list(model.parameters())

[Parameter containing:
 tensor([[ 0.0933,  0.5745, -0.2043],
         [ 0.3817, -0.3547,  0.0262]], requires_grad=True),
 Parameter containing:
 tensor([-0.4140,  0.1716], requires_grad=True)]

We can use the model to generate predictions in the exact same way as before:

In [38]:
# Generate predictions
preds = model(inputs)
preds

tensor([[ 36.0982,   5.3973],
        [ 45.5504,   5.3693],
        [ 72.8286, -12.6311],
        [ 26.2410,  24.8225],
        [ 46.8689,  -5.7086],
        [ 36.0982,   5.3973],
        [ 45.5504,   5.3693],
        [ 72.8286, -12.6311],
        [ 26.2410,  24.8225],
        [ 46.8689,  -5.7086],
        [ 36.0982,   5.3973],
        [ 45.5504,   5.3693],
        [ 72.8286, -12.6311],
        [ 26.2410,  24.8225],
        [ 46.8689,  -5.7086]], grad_fn=<AddmmBackward>)

## Loss Function

Instead of defining a loss function manually, we can use the built-in loss function `mse_loss`.

In [39]:
# Import nn.functional
import torch.nn.functional as F

The `nn.functional` package contains many useful loss functions and several other utilities. 

In [40]:
# Define loss function
loss_fn = F.mse_loss

Let's compute the loss for the current predictions of our model.

In [41]:
loss = loss_fn(model(inputs), targets)
print(loss)

tensor(5718.0928, grad_fn=<MseLossBackward>)


## Optimizer

Instead of manually manipulating the model's weights & biases using gradients, we can use the optimizer `optim.SGD`. SGD stands for `stochastic gradient descent`. It is called `stochastic` because samples are selected in batches (often with random shuffling) instead of as a single group.

In [42]:
# Define optimizer
opt = torch.optim.SGD(model.parameters(), lr=1e-5)

Note that `model.parameters()` is passed as an argument to `optim.SGD`, so that the optimizer knows which matrices should be modified during the update step. Also, we can specify a learning rate which controls the amount by which the parameters are modified.

## Train the model

We are now ready to train the model. We'll follow the exact same process to implement gradient descent:

1. Generate predictions

2. Calculate the loss

3. Compute gradients w.r.t the weights and biases

4. Adjust the weights by subtracting a small quantity proportional to the gradient

5. Reset the gradients to zero

The only change is that we'll work batches of data, instead of processing the entire training data in every iteration. Let's define a utility function `fit` which trains the model for a given number of epochs.

In [43]:
# Utility function to train the model
def fit(num_epochs, model, loss_fn, opt):
    
    # Repeat for given number of epochs
    for epoch in range(num_epochs):
        
        # Train with batches of data
        for xb,yb in train_dl:
            
            # 1. Generate predictions
            pred = model(xb)
            
            # 2. Calculate loss
            loss = loss_fn(pred, yb)
            
            # 3. Compute gradients
            loss.backward()
            
            # 4. Update parameters using gradients
            opt.step()
            
            # 5. Reset the gradients to zero
            opt.zero_grad()
        
        # Print the progress
        if (epoch+1) % 10 == 0:
            print('Epoch [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, loss.item()))

Some things to note above:

* We use the data loader defined earlier to get batches of data for every iteration.

* Instead of updating parameters (weights and biases) manually, we use `opt.step` to perform the update, and `opt.zero_grad` to reset the gradients to zero.

* We've also added a log statement which prints the loss from the last batch of data for every 10th epoch, to track the progress of training. `loss.item` returns the actual value stored in the loss tensor.

Let's train the model for 100 epochs.

In [44]:
fit(100, model, loss_fn, opt)

Epoch [10/100], Loss: 917.0132
Epoch [20/100], Loss: 296.0120
Epoch [30/100], Loss: 195.5623
Epoch [40/100], Loss: 139.5297
Epoch [50/100], Loss: 139.4575
Epoch [60/100], Loss: 96.9141
Epoch [70/100], Loss: 101.5640
Epoch [80/100], Loss: 89.2382
Epoch [90/100], Loss: 50.3481
Epoch [100/100], Loss: 58.9096


Let's generate predictions using our model and verify that they're close to our targets.

In [45]:
# Generate predictions
preds = model(inputs)
preds

tensor([[ 58.0900,  72.1446],
        [ 78.2882,  98.9743],
        [124.7621, 133.4965],
        [ 27.4780,  47.5143],
        [ 91.3867, 110.1211],
        [ 58.0900,  72.1446],
        [ 78.2882,  98.9743],
        [124.7621, 133.4965],
        [ 27.4780,  47.5143],
        [ 91.3867, 110.1211],
        [ 58.0900,  72.1446],
        [ 78.2882,  98.9743],
        [124.7621, 133.4965],
        [ 27.4780,  47.5143],
        [ 91.3867, 110.1211]], grad_fn=<AddmmBackward>)

In [46]:
# Compare with targets
targets

tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])

Indeed, the predictions are quite close to our targets, and now we have a fairly good model to predict crop yields for apples and oranges by looking at the average temperature, rainfall and humidity in a region.

## Commit and update the notebook

As a final step, we can record a new version of the notebook using the `jovian` library.

In [93]:
import jovian

In [95]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..


UnboundLocalError: local variable 'username' referenced before assignment

Note that running `jovian.commit` a second time records a new version of your existing notebook. With Jovian.ml, you can avoid creating copies of your Jupyter notebooks and keep versions organized. Jovian also provides a visual diff ([example](https://jovian.ml/aakashns/keras-mnist-jovian/diff?base=8&remote=2)) so you can inspect what has changed between different versions:

![jovian-commenting](https://i.imgur.com/HF1cOVt.png)

## Further Reading

We've covered a lot of ground this this tutorial, including *linear regression* and the *gradient descent* optimization algorithm. Here are a few resources if you'd like to dig deeper into these topics:

* For a more detailed explanation of derivates and gradient descent, see [these notes from a Udacity course](https://storage.googleapis.com/supplemental_media/udacityu/315142919/Gradient%20Descent.pdf). 

* For an animated visualization of how linear regression works, [see this post](https://hackernoon.com/visualizing-linear-regression-with-pytorch-9261f49edb09).

* For a more mathematical treatment of matrix calculus, linear regression and gradient descent, you should check out [Andrew Ng's excellent course notes](https://github.com/Cleo-Stanford-CS/CS229_Notes/blob/master/lectures/cs229-notes1.pdf) from CS229 at Stanford University.

* To practice and test your skills, you can participate in the [Boston Housing Price Prediction](https://www.kaggle.com/c/boston-housing) competition on Kaggle, a website that hosts data science competitions.

With this, we complete our discussion of linear regression in PyTorch, and we’re ready to move on to the next topic: *Logistic regression*.

# Trying out myself

## Manual Method 

yield_apple  = w11 * temp + w12 * rainfall + w13 * humidity + b1

yield_orange = w21 * temp + w22 * rainfall + w23 * humidity + b2

In [20]:
import numpy as np
import torch

In [27]:
inputs = np.array([[73, 67, 43],[91, 88,  64], [87, 134, 58], [102, 43, 37], [69, 96, 70]], dtype='float32')
targets = np.array([[56, 70],[81, 101], [119, 133], [22, 37], [103, 119]], dtype='float32')

In [28]:
inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)
print(inputs)
print(targets)

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.]])
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


In [31]:
weights = torch.randn(2,3, requires_grad=True) #need require grad coz its grad function
bias = torch.randn(2, requires_grad=True)
print(weights)
print(bias)

tensor([[ 0.0893,  0.7641, -0.5488],
        [ 0.4831,  0.0727,  0.2150]], requires_grad=True)
tensor([-0.1454,  0.5987], requires_grad=True)


In [32]:
def model(x):
    return x @ weights.T + bias
    

In [35]:
def mse(a, b):
    diff = a- b
    return torch.sum(diff * diff)/diff.numel()

tensor([[33.9731, 49.9823],
        [40.1024, 64.7206],
        [78.1886, 64.8443],
        [21.5170, 60.9554],
        [40.9575, 55.9651]], grad_fn=<AddBackward0>)
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


In [41]:
learningRate = 1e-5
print(targets)
for i in range(100):
    prediction = model(inputs)
    loss = mse(prediction, targets)
    loss.backward()
    with torch.no_grad():
        weights -= weights.grad * learningRate
        bias -= bias.grad * learningRate
        weights.grad.zero_()
        bias.grad.zero_()
    if i%10 == 0:
        print(prediction)


tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])
tensor([[39.0010, 55.2431],
        [46.7321, 71.6676],
        [86.1089, 73.2819],
        [26.3947, 65.9219],
        [47.3937, 62.7847]], grad_fn=<AddBackward0>)
tensor([[ 59.5218,  76.5307],
        [ 73.9418, 100.0042],
        [119.1642, 109.1811],
        [ 45.4644,  84.2400],
        [ 74.3373,  91.6465]], grad_fn=<AddBackward0>)
tensor([[ 61.3157,  78.1647],
        [ 76.5120, 102.4627],
        [122.9518, 114.1291],
        [ 46.0922,  83.4228],
        [ 77.5357,  95.4470]], grad_fn=<AddBackward0>)
tensor([[ 61.3712,  77.9861],
        [ 76.7894, 102.5224],
        [123.9711, 116.1105],
        [ 45.0587,  80.9505],
        [ 78.4954,  96.8608]], grad_fn=<AddBackward0>)
tensor([[ 61.1968,  77.5803],
        [ 76.7564, 102.2675],
        [124.5682, 117.5969],
        [ 43.8615,  78.3926],
        [ 79.1192,  97.8916]], grad_fn=<AddBackward0>)
tensor([[ 61.0011,  77.1

## using Pytorch builtin

In [72]:
import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import TensorDataset
from torch.utils.data import DataLoader
import torch.nn.functional as F

In [73]:
inputs = np.array([[73, 67, 43],[91, 88,  64], [87, 134, 58], [102, 43, 37], [69, 96, 70], [73, 67, 43],[91, 88,  64], [87, 134, 58], [102, 43, 37], [69, 96, 70]], dtype='float32')
targets = np.array([[56, 70],[81, 101], [119, 133], [22, 37], [103, 119], [56, 70],[81, 101], [119, 133], [22, 37], [103, 119]], dtype='float32')

In [74]:
inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)
print(inputs)
print(targets)

tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.],
        [ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.]])
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


In [75]:
#create tensor dataset
trainingset = TensorDataset(inputs, targets)
print(trainingset[:])

(tensor([[ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.],
        [ 73.,  67.,  43.],
        [ 91.,  88.,  64.],
        [ 87., 134.,  58.],
        [102.,  43.,  37.],
        [ 69.,  96.,  70.]]), tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]]))


In [76]:
# create data loader
batch_size = 2
trainingset_loaded = DataLoader(trainingset, batch_size, shuffle=True)

In [77]:
for x, y in trainingset_loaded:
    print(x)
    print(y)

tensor([[ 87., 134.,  58.],
        [ 69.,  96.,  70.]])
tensor([[119., 133.],
        [103., 119.]])
tensor([[ 73.,  67.,  43.],
        [ 87., 134.,  58.]])
tensor([[ 56.,  70.],
        [119., 133.]])
tensor([[91., 88., 64.],
        [69., 96., 70.]])
tensor([[ 81., 101.],
        [103., 119.]])
tensor([[ 91.,  88.,  64.],
        [102.,  43.,  37.]])
tensor([[ 81., 101.],
        [ 22.,  37.]])
tensor([[102.,  43.,  37.],
        [ 73.,  67.,  43.]])
tensor([[22., 37.],
        [56., 70.]])


In [71]:
#init weights and bias through (create model)
model = nn.Linear(3, 2)  # 3 - no of variable. 2 - no of targets
print(model.weight) 
print(model.bias)
list(model.parameters())

Parameter containing:
tensor([[ 0.4481, -0.2487, -0.3269],
        [-0.5548, -0.2367, -0.3127]], requires_grad=True)
Parameter containing:
tensor([0.3015, 0.1358], requires_grad=True)


[Parameter containing:
 tensor([[ 0.4481, -0.2487, -0.3269],
         [-0.5548, -0.2367, -0.3127]], requires_grad=True),
 Parameter containing:
 tensor([0.3015, 0.1358], requires_grad=True)]

In [81]:
#define optimiser (automatically steps) - SGD stands for stochastic gradient descent
opt = torch.optim.SGD(model.parameters(), lr=1e-5)  #modifies weight and bias with learning rate

In [88]:
def fit(num_epochs):
    for epoch in range(num_epochs):
        for x, y in trainingset_loaded:
            pred = model(x)
            loss = F.mse_loss(pred, y)
            loss.backward()
            opt.step()
            opt.zero_grad()
            
        if (epoch+1) % 10 == 0:
            print('Epoch [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, loss.item()))

In [89]:
fit(100)

Epoch [10/100], Loss: 5.0645
Epoch [20/100], Loss: 4.2333
Epoch [30/100], Loss: 18.7367
Epoch [40/100], Loss: 13.4372
Epoch [50/100], Loss: 2.5913
Epoch [60/100], Loss: 2.4435
Epoch [70/100], Loss: 2.4055
Epoch [80/100], Loss: 12.7101
Epoch [90/100], Loss: 9.6091
Epoch [100/100], Loss: 1.4385


In [87]:
preds = model(inputs)
print(preds)
print(targets)

tensor([[ 58.0146,  70.9081],
        [ 79.7079,  98.1245],
        [123.8315, 138.4769],
        [ 24.7679,  39.5118],
        [ 95.4214, 113.2306],
        [ 58.0146,  70.9081],
        [ 79.7079,  98.1245],
        [123.8315, 138.4769],
        [ 24.7679,  39.5118],
        [ 95.4214, 113.2306]], grad_fn=<AddmmBackward>)
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])


In [98]:
import jovian

In [101]:
jovian.commit()


<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..


UnboundLocalError: local variable 'username' referenced before assignment